### PR TITLE
Save "Sequences" and fire them via a web call

### PR DESCRIPTION
**Pull request by calebcauthon**
#### Cree "Tree-Style" sequences and fire them:
image

#### Save sequences to the DB
image

#### And fire different sequences via a web call
image

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -21,4 +21,9 @@
     t.text "step_guid"
   end
 
+  create_table "sequences", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
+    t.string "name"
+    t.string "steps"
+  end
```
Comment (calebcauthon) [inline db/schema.rb]: New table...

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -9,13 +9,29 @@ def all
...
+  end
+
+  def fire
+    sequence = Sequence.where(name: params[:message]).first
+
+    return render plain: "no sequence found for \"#{params[:message]}\"" unless sequence
+
+    steps = JSON.parse(sequence.steps, symbolize_names: true)
+    steps.each { |s| s[:config] = { name: s[:name] } }
+    steps.unshift({ type: 'WorkflowInitiator' })
```
Comment (calebcauthon) [inline app/controllers/events_controller.rb]: Here we grab a sequence from the DB, add a beginning "web initiator event" and then make a workflow from that and fire the whole thing.

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -9,13 +9,29 @@ def all
...
     originating_data = { message: params[:message] }
-    steps = JSON.parse(request.body.read, symbolize_names: true)
+    steps_encoded = request.body.read
+    steps = JSON.parse(steps_encoded, symbolize_names: true)
 
-    workflow = Workflow.build(steps: steps)
+    sequence = Sequence.new
+    sequence.name = steps[0][:name]
+    sequence.steps = steps_encoded
+    sequence.save
```
Comment (calebcauthon) [inline app/controllers/events_controller.rb]: Rather than fire here, we just save the sequence to the db.

**Comment by calebcauthon** [comment]
Not intending this to necessarily go through, but I just wanted to see it happen. Was just itching to start putting pieces together. But let me know what you think anyway.

Also just changed some layout for fun, which I meant to separate out but forgot to.

**Comment by calebcauthon** [comment]
is this a way to save a workflow and have it fired elsewhere- yep!

**Comment by calebcauthon** [comment]
yes, jumping the gun. on purpose.

**Comment by calebcauthon** [comment]
I made this issue to keep our terminology straight https://github.com/darrencauthon/fasten/issues/27

**Comment by calebcauthon** [comment]
I'm going to address the 'tree' issue with this first.

File: app/models/sequence.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+class Sequence < ApplicationRecord
+end
```
Comment (darrencauthon) [inline app/models/sequence.rb]: Kill this, and add ``` < ApplicationRecord``` to Workflow.

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -9,26 +9,23 @@ def all
...
-
-    workflow = Workflow.build(steps: steps)
+    steps_encoded = request.body.read
+    steps = JSON.parse(steps_encoded, symbolize_names: true)
 
-    result = workflow.start(originating_data)
+    workflow = Workflow.new
+    workflow.name = steps[:first_step][:name]
+    workflow.steps_encoded_as_json = steps_encoded
+    workflow.save
```
Comment (calebcauthon) [inline app/controllers/events_controller.rb]: Here's the essence of making a new workflow

File: app/models/workflow.rb [hunk 6]
```diff
@@ -1,33 +1,29 @@
...
-      .each_with_index { |x, i| x[:next_steps] = [definition[:steps][i+1]].reject { |x| x.nil? } }
-      .each_with_index { |x, i| x[:method] = lambda { |e| Workflow.build_event_handler_for(x).receive e } }
+    saved_workflow.steps = JSON.parse(saved_workflow.steps_encoded_as_json, symbolize_names: true)
+    saved_workflow.build
 
-    workflow.first_step = definition[:steps].first
-    workflow.first_step[:method] = lambda { |e| Workflow.build_event_handler_for(workflow.first_step).fire e }
-
-    workflow
+    saved_workflow
```
Comment (calebcauthon) [inline app/models/workflow.rb]: Here's the function where we encapsulate all that 'build a workflow' stuff.

**Comment by calebcauthon** [comment]
back at ya for review

**Comment by calebcauthon** [comment]
@darrencauthon I'm closing this, I'll bring it up again in a cleaner slimmer way after you get your 'spark' changes in.

**Pull request closed without merging**